<h3>To accurately predict Insurance costs :- </h3>
(for given dataset)

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
df=pd.read_csv("insurance.csv")
df

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

# Preprocessing

In [3]:
df["sex"]=df["sex"].replace({"female":0,"male":1})
df["smoker"]=df["smoker"].replace({"no":0,"yes":1})

In [4]:
x=df.iloc[:,:-1]
y=df["charges"].values

In [5]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
col_trans=ColumnTransformer(transformers=[('encoder',OneHotEncoder(),[5])],remainder='passthrough')
x=np.array(col_trans.fit_transform(x))

In [7]:
y=y.reshape((-1,1))

# Division in Train & test

In [8]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1)

## Multiple Linear Regression

In [9]:
from sklearn.linear_model import LinearRegression
reg=LinearRegression()
reg.fit(x_train,y_train)
y_pred=reg.predict(x_test)

In [10]:
from sklearn.metrics import r2_score
print("Result of Multiple Linear Regression Model on data is : ",r2_score(y_test,y_pred))

Result of Multiple Linear Regression Model on data is :  0.7623311844057113


# SVR Model

In [11]:
from sklearn.preprocessing import StandardScaler
ss=StandardScaler()
ss1=StandardScaler()
sd=StandardScaler()
sd1=StandardScaler()
x_std=ss.fit_transform(x_train)
y_std=sd.fit_transform(y_train)
x_test_std=ss.transform(x_test)

from sklearn.svm import SVR
reg_svr=SVR(kernel='rbf')
reg.fit(x_std,y_std)
y_pred_std=reg.predict(x_test_std)
print("Result of SVR Model on data is : ",r2_score(y_test,sd.inverse_transform(y_pred_std)))

Result of SVR Model on data is :  0.7623258010962115


# Decision Tree

In [12]:
from sklearn.tree import DecisionTreeRegressor
reg_tree=DecisionTreeRegressor(random_state=1)
reg_tree.fit(x_train,y_train)
y_pred_d_tree=reg_tree.predict(x_test)
print("Result of Decision tree Model on data is : ",r2_score(y_test,y_pred_d_tree))

Result of Decision tree Model on data is :  0.7510687791159403


# Random Forest Reg

In [13]:
from sklearn.ensemble import RandomForestRegressor
reg_ran=RandomForestRegressor(n_estimators=300,random_state=0)
y_train_1d=np.ravel(y_train)
reg_ran.fit(x_train,y_train_1d)
y_pred_r_forest=reg_ran.predict(x_test)
print("Result of Random Forest Model on data is : ",r2_score(y_test,y_pred_r_forest))

Result of Random Forest Model on data is :  0.8560430320860553



<h4>Thus it can be seen that "Random Forest Regression Model" has the best fit to out dataset :</h4>
<b>i.e R square value = 0.856043</b>